stdlib/public/core/DictionaryBuilder.swift

```diff
@@ -42,3 +42,85 @@ struct _DictionaryBuilder<Key: Hashable, Value> {
+  /// instance without allocating a pair of intermediary buffers.  Pass the
+  /// required capacity and a closure that can intialize the dictionary's
+  /// elements. The closure must return `c`, the number of initialized elements
+  /// in both buffers, such that the elements in the range `0..<c` are
+  /// initialized and the elements in the range `c..<capacity` are
+  /// uninitialized. The resulting dictionary has a `count` equal to `c`.
+  ///
+  /// The closure must not add any duplicate keys to the keys buffer.  The
+  /// buffers passed to the closure are only valid for the duration of the call.
+  /// After the closure returns, this initializer moves all initialized elements
+  /// into their correct buckets.
+  ///
+  /// - Parameters:
+  ///   - capacity: The capacity of the new dictionary.
+  ///   - body: A closure that can initialize the dictionary's elements. This
+  ///     closure must return the count of the initialized elements, starting at
+  ///     the beginning of the buffer.
+  @inlinable
+  public // SPI(Foundation)
+  init(
+    _unsafeUninitializedCapacity capacity: Int,
+    initializingWith initializer: (
+      _ keys: UnsafeMutableBufferPointer<Key>,
+      _ values: UnsafeMutableBufferPointer<Value>,
+      _ initializedCount: inout Int
+    ) -> Void
+  ) {
+    let native = _NativeDictionary<Key, Value>(capacity: capacity)
+    let keys = UnsafeMutableBufferPointer<Key>(
+      start: native._keys,
+      count: capacity)
+    let values = UnsafeMutableBufferPointer<Value>(
+      start: native._values,
+      count: capacity)
+    var count = 0
+    initializer(keys, values, &count)
+    _precondition(count >= 0 && count <= capacity)
+
+    // Hash elements into the correct buckets.
+    var bucket = _HashTable.Bucket(offset: 0)
+    while bucket.offset < count {
+      if native.hashTable._isOccupied(bucket) {
+        // We've moved an element here in a previous iteration.
+        bucket.offset += 1
+        continue
+      }
+      let hashValue = native.hashValue(for: native._keys[bucket.offset])
+      let target: _HashTable.Bucket
+      if _isDebugAssertConfiguration() {
+        let (b, found) = native.find(
+          native._keys[bucket.offset],
+          hashValue: hashValue)
+        guard !found else {
+          _preconditionFailure("Duplicate keys found")
+        }
+        native.hashTable.insert(b)
+        target = b
+      } else {
+        target = native.hashTable.insertNew(hashValue: hashValue)
+      }
+      if target < bucket || target.offset >= count {
+        // Target bucket is uninitialized
+        native.moveEntry(from: bucket, to: target)
+        bucket.offset += 1
+      } else if target == bucket {
+        // Already in place
+        bucket.offset += 1
+      } else {
+        // Swap into place, then try again with the swapped-in value.
+        native.swapEntry(target, with: bucket)
+      }
```

Comment:
This is a cool algorithm, @lorentey 👏

Reply:
I had to work it out myself because I couldn't picture how to hash-in-place like this, but I think I got it. Nice work. (Did you adapt it from somewhere or come up with it yourself?)

Does this obsolete DictionaryBuilder? 

For my own understanding:
- The storage is separated into three regions: "before `bucket`", "`bucket` up to `count`", and "`count` and after".
- Everything in "before `bucket`" and "`count` and after" is either uninitialized or in use. Everything in "`bucket` up to `count`" is either unprocessed or in use.
- "in use" is tracked by the bitmap in `native.hashTable`, the same way it would be for a working Dictionary. Whenever we see such a bucket in the middle section, we skip over it, and *`native.hashTable.insertNew` will never return an in-use bucket*. (This is the key thing I was missing when trying to picture the algorithm on my own.)
- When processing an item, we're basically deciding how to move it from the middle section. Either we're moving it elsewhere in the middle section, in which case we do a swap so that we don't lose the unprocessed item that was there, or we're moving it outside the middle section, in which case it's okay to leave behind an uninitialized hole (because we're about to advance `bucket`).

Reply:
That's a nice analysis & it's exactly right -- in fact, I'll steal it to document the invariants!

This is unlikely to be a new algorithm -- I can swear I saw something like this done before, although maybe not in these exact circumstances. (I'll assume 

We're experimenting with switching some forms of bridging from `_DictionaryBuilder` to this, but it's only an experiment at this point. This is designed specifically to work with the narrow use case of `-[NSDictionary objects:keys:count:]`. 

It's quite more fiddly than `_DictionaryBuilder`, so if the input comes from a sequence of unique key/value pairs, then I expect the builder will likely still be faster. (Although benchmarks may surprise us!)

Reply:
I wonder if there's a tiny simplification in going backwards instead of forwards—indeed, @Catfish-Man described it to me that way at first. That way you only have two regions to think about (and for a concrete benefit, only have to make one comparison).

Reply:
Good idea; applied!